This notebook is sourced from our notes here - https://telestreak.com/topics/tech/data/exploration/

Welcome to Lesson-1 of the **Data Exploration with Python** series!
In this notebook, we will explore a public dataset that contains Country/Mission wise [OCI details](https://data.gov.in/resources-from-web-service/3670801) from the Indian Open Government Data Platform. I am using some of these less explored datasets mimicing a typical work situation where the given data is probably from a new domain or has information that you may not have observed before.

We will cover the lesson through below steps, introducing you to some basic commands/functions to analyze the dataset:
1. Import required python libraries
2. Read the dataset into a dataframe (since it is small file, the dataset already saved to my github repository in /telestreak/data_exp_setup/datasets)
3. Analyze the dataset through various commands/functions

Every _cell_ below is either a **markdown** cell - containing formatted text OR a **code** cell - containing commands that can be executed.
To execute a cell, place the cursor within the cell and hit the _play_ button from the toolbar at the top of this notebook or use the keyboard shortcut _Shift + Enter (or Return)_

## 1. Import required Python libararies 

In [16]:
import pandas as pd

## 2. Read the dataset into a _dataframe_ 

In [17]:
# Read the .csv file that is saved in this repository, into a dataframe called "df"
df = pd.read_csv('../datasets/oci_dataset_2013.csv')

A _dataframe_ is a two dimensional (consisting of rows and columns), tabular data structure provided by Python's _pandas_ library. Below command shows that _df_ is of type DataFrame from pandas library.

In [18]:
type(df)

pandas.core.frame.DataFrame

In [19]:
df.shape

(73575, 9)

In [20]:
df.columns

Index(['Country', 'Mission', 'Date', 'OCI_Registered', 'OCI_Issued\n\n',
       'Image_Scanned ', 'OCI_Granted', 'OCI_Despatched_to_mission',
       'OCI_Enquiries'],
      dtype='object')

There are 73575 rows/records in the dataset with 9 columns/attributes. Let's display a sample of the file we just read from the .csv.

In [21]:
df.head()

Country    Mission        Date  OCI_Registered  OCI_Issued\n\n  \
0  AUSTRALIA  MELBOURNE  01-01-2013             0.0             1.0   
1  AUSTRALIA      PERTH  01-01-2013             0.0            22.0   
2    AUSTRIA     VIENNA  01-01-2013             0.0             1.0   
3     CANADA     OTTAWA  01-01-2013             0.0            11.0   
4     CANADA    TORONTO  01-01-2013             0.0            49.0   

   Image_Scanned   OCI_Granted  OCI_Despatched_to_mission  OCI_Enquiries  
0             0.0          0.0                        0.0            0.0  
1             0.0          0.0                        0.0            0.0  
2             0.0          0.0                        0.0            0.0  
3             0.0         11.0                        0.0            0.0  
4            47.0          0.0                        0.0            0.0

Looks like one of the column names - "OCI_Issued" has new line characters (\n). Let's give it a proper name.
To learn more about the parameters that go in any function call, place your cursor on the function name and hit _Shift + Tab_ in your jupyter notebook.

In [22]:
df.rename(columns={'OCI_Issued\n\n':'OCI_Issued'}, inplace=True)

Setting _inplace_ option to True, makes the changes to the dataframe within the same command, **in place** without creating a new dataframe with the modifications.

In [23]:
df.head()

Country    Mission        Date  OCI_Registered  OCI_Issued  \
0  AUSTRALIA  MELBOURNE  01-01-2013             0.0         1.0   
1  AUSTRALIA      PERTH  01-01-2013             0.0        22.0   
2    AUSTRIA     VIENNA  01-01-2013             0.0         1.0   
3     CANADA     OTTAWA  01-01-2013             0.0        11.0   
4     CANADA    TORONTO  01-01-2013             0.0        49.0   

   Image_Scanned   OCI_Granted  OCI_Despatched_to_mission  OCI_Enquiries  
0             0.0          0.0                        0.0            0.0  
1             0.0          0.0                        0.0            0.0  
2             0.0          0.0                        0.0            0.0  
3             0.0         11.0                        0.0            0.0  
4            47.0          0.0                        0.0            0.0

In [24]:
# this column is of object type. Let's convert this to date format.
df['Date'].head()

0    01-01-2013
1    01-01-2013
2    01-01-2013
3    01-01-2013
4    01-01-2013
Name: Date, dtype: object

In [25]:
df['Date'] = pd.to_datetime(df['Date'])

## 3. Analyze dataset 

In [26]:
df['Date'].min()

Timestamp('2013-01-01 00:00:00')

Fits the description for the [dataset](https://data.gov.in/resources-from-web-service/3670801), that the OCI details starts from 2013-01-01 and is available up to 2018-12-02.

In [27]:
df['Date'].max()

Timestamp('2018-12-02 00:00:00')

Does it have all dates between start and end?
[Stackoverflow reference](https://stackoverflow.com/questions/52044348/check-for-any-missing-dates-in-the-index) for the command below.

In [28]:
pd.date_range(start='2013-01-01', end='2018-12-02').difference(df['Date'])

DatetimeIndex(['2013-02-24', '2013-03-11', '2013-04-14', '2013-05-05',
               '2013-07-07', '2013-10-13', '2013-10-27', '2013-11-17',
               '2013-11-24', '2013-12-14',
               ...
               '2018-11-21', '2018-11-22', '2018-11-23', '2018-11-24',
               '2018-11-25', '2018-11-26', '2018-11-27', '2018-11-28',
               '2018-11-29', '2018-11-30'],
              dtype='datetime64[ns]', length=351, freq=None)

Try changing the _start_ and _end_ dates above to one year at a time, you'll notice that 2018 has numerous missing days, followed by 2014. For other years, there are some random days that are missing.
I am assuming these are days when there was no activity in these missions.

A handy command to generate some basic stats about the data

In [29]:
df.describe()

OCI_Registered    OCI_Issued  Image_Scanned    OCI_Granted  \
count    73525.000000  73525.000000    73525.000000  73525.000000   
mean        21.001265     21.054104       19.315607     19.816770   
std         45.195086     53.005048       49.512888     58.327617   
min          0.000000      0.000000        0.000000      0.000000   
25%          1.000000      0.000000        0.000000      0.000000   
50%          4.000000      3.000000        0.000000      0.000000   
75%         20.000000     17.000000       16.000000     10.000000   
max       1427.000000   1062.000000     1082.000000   1655.000000   

       OCI_Despatched_to_mission  OCI_Enquiries  
count               73525.000000   73525.000000  
mean                   45.726134      24.916695  
std                   117.374456      94.417887  
min                     0.000000       0.000000  
25%                     0.000000       0.000000  
50%                     6.000000       1.000000  
75%                    37.000000      17.000000  
max                  1984.000000    4715.000000

Looks like the max number of OCIs registered in a day is 1427. Let's check the country and mission that has the highest OCIs registered in a day.

In [30]:
df[df['OCI_Registered'] == 1427]

Country  Mission       Date  OCI_Registered  OCI_Issued  \
57106  UNITED STATES  CHICAGO 2017-02-03          1427.0        63.0   

       Image_Scanned   OCI_Granted  OCI_Despatched_to_mission  OCI_Enquiries  
57106            12.0         27.0                       47.0           59.0

In [31]:
# number of countries in the dataset
df['Country'].nunique()

119

In [32]:
# number of missions
df['Mission'].nunique()

185

20 countries regsitering most OCIs. Here I am grouping by _Country_, to get total OCI registrations. I then sort the values from the result in descending order

In [33]:
df.groupby('Country')['OCI_Registered'].sum().sort_values(ascending=False).head(20)

Country
UNITED STATES     432168.0
UNITED KINGDOM    420678.0
AUSTRALIA         191403.0
CANADA             97061.0
INDIA              79840.0
MALAYSIA           37112.0
SINGAPORE          35958.0
ITALY              31434.0
GERMANY            23789.0
NEW ZEALAND        22383.0
FRANCE             21291.0
KENYA              13619.0
IRELAND            13541.0
NETHERLANDS        10095.0
UAE                 7899.0
TANZANIA            6937.0
SOUTH AFRICA        6438.0
NORWAY              6378.0
AUSTRIA             5769.0
BELGIUM             5396.0
Name: OCI_Registered, dtype: float64

In [34]:
# 20 countries with few OCIs registered
df.groupby('Country')['OCI_Registered'].sum().sort_values(ascending=True).head(20)

Country
MALI                         2.0
CUBA                         3.0
SOUTH SUDAN                  5.0
SERBIA                       5.0
ALGERIA                      6.0
MONGOLIA                     8.0
NIB                         12.0
LEBANON                     13.0
BANGLADESH                  14.0
TUNISIA                     17.0
NEPAL                       18.0
TAJIKISTAN                  18.0
SNG                         18.0
NIA                         21.0
GUATEMALA                   21.0
MOROCCO                     23.0
ARMENIA                     25.0
NAMIBIA                     26.0
ZAIRE (DEM REP OF CONGO)    28.0
KYRGYZSTAN                  29.0
Name: OCI_Registered, dtype: float64

In [35]:
# 20 missions regsitering most OCIs
df.groupby('Mission')['OCI_Registered'].sum().sort_values(ascending=False).head(20)

Mission
LONDON           291260.0
NEW YORK CGI     145861.0
BIRMINGHAM       122445.0
SAN FRANCISCO     88126.0
SYDNEY            72643.0
CHICAGO           67646.0
MELBOURNE         66003.0
TORONTO           48918.0
HOUSTON           46979.0
ATLANTA           46590.0
VANCOUVER         42512.0
KUALA LUMPUR      37112.0
WASHINGTON DC     36966.0
SINGAPORE         35958.0
FRRO MUMBAI       27711.0
CANBERRA          27398.0
PERTH             25359.0
WELLINGTON        22383.0
PARIS             21291.0
MILAN             21055.0
Name: OCI_Registered, dtype: float64

In [ ]:
# Can you complete this command to find the missions with most enqueries?
df.groupby('')[''].sum().sort_values(ascending=False).head(20)

Show a sample set with countries having _UNITED_ in their name.

In [36]:
df[df['Country'].str.contains('UNITED', na=True)].head()

Country        Mission       Date  OCI_Registered  OCI_Issued  \
23  UNITED KINGDOM         LONDON 2013-01-01             0.0       207.0   
24   UNITED STATES        CHICAGO 2013-01-01            47.0         0.0   
25   UNITED STATES   NEW YORK CGI 2013-01-01            75.0       181.0   
26   UNITED STATES  SAN FRANCISCO 2013-01-01            54.0       127.0   
27   UNITED STATES  WASHINGTON DC 2013-01-01            12.0        84.0   

    Image_Scanned   OCI_Granted  OCI_Despatched_to_mission  OCI_Enquiries  
23             0.0          0.0                        0.0            0.0  
24            51.0          0.0                        0.0            0.0  
25            42.0        126.0                        0.0            0.0  
26            55.0         97.0                        0.0            0.0  
27             4.0         83.0                        0.0            0.0

What is the mean OCI registrations in Washington DC by day?

In [37]:
df[df['Mission'] == 'WASHINGTON DC']['OCI_Registered'].mean()

27.443207126948774

In [38]:
df['OCI_Registered'].sum()

1544118.0

In [39]:
df['OCI_Issued'].sum()

1548003.0

In [40]:
df['OCI_Granted'].sum()

1457028.0

OCI registered in a given day doesn't necessarily get issued or granted the same day. But the total number of OCIs issued being higher than the total number of OCIs registered for this time period, could indicate:
1. The additional OCIs issued are from a past period, not in this data set
2. There are times when people have to get their OCIs re-issued (for example, when they reach the age of 19). Assuming they don't have to register again, the number of issued could be high. However, the difference is just ~4K over ~1.5MM - making (1) a more reasonable explanation.

## Summary

In this lesson, we used some basic commands and functions from the _pandas_ library to read a public dataset having OCI details. We also performed some basic data transformations and analysis to build familiarity on this dataset.

Data transformations:
- renaming a dataframe column
- changing datatype of an _object_ type to _date_

Functions for data exploration:
- read_csv()
- shape()
- columns()
- head()
- min(), max(), sum(), mean(), describe()
- date_range()
- nunique()
- groupby()
- sort_values()